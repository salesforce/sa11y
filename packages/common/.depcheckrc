# Ignore webdriverio - we need the types for `WebdriverIO.Browser`
ignores: ['webdriverio']
