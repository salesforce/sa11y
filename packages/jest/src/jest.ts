/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { assertAccessible } from '@sa11y/assert';
import { A11yConfig, extended } from '@sa11y/preset-rules';
import { matcherHint } from 'jest-matcher-utils';

export const matcherHintMsg = `expected document to have no accessibility violations but found following issues: `;

// Type def for custom jest a11y matcher toBeAccessible
// Ref: https://jestjs.io/docs/en/expect.html#expectextendmatchers
declare global {
    // eslint-disable-next-line @typescript-eslint/no-namespace
    namespace jest {
        interface Matchers<R> {
            toBeAccessible(config?: A11yConfig): Promise<CustomMatcherResult>;
            toBeAccessibleWith(config: A11yConfig): Promise<CustomMatcherResult>;
        }
    }
}

/**
 * Jest expect matcher to check DOM for accessibility issues
 * @param receivedDom - DOM to be tested for accessibility
 * @param config - A11yConfig to be used to test for accessibility. Defaults to extended.
 */
export async function toBeAccessible(
    receivedDom: Document = document,
    config: A11yConfig = extended
): Promise<jest.CustomMatcherResult> {
    let isAccessible = true;
    let a11yViolations = '';

    try {
        await assertAccessible(receivedDom, config);
    } catch (e) {
        isAccessible = false;
        a11yViolations = e;
    }
    return {
        pass: isAccessible,
        // Display assertion for the report when a test fails
        message: (): string => matcherHint(`${toBeAccessible.name}: ${matcherHintMsg} \n\n ${a11yViolations}`),
    };
}

/**
 * Jest expect matcher to check DOM for accessibility issues with a `@sa11y/preset-rule`
 * @param receivedDom - DOM to be tested for accessibility
 * @param config - A11yConfig to be used to test for accessibility.
 */
export function toBeAccessibleWith(
    receivedDom: Document = document,
    config: A11yConfig
): Promise<jest.CustomMatcherResult> {
    return toBeAccessible(receivedDom, config);
}
