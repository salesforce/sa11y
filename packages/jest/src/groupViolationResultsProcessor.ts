/*
 * Copyright (c) 2023, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { AggregatedResult, AssertionResult, TestResult } from '@jest/test-result';
import { log } from '@sa11y/common';
import { A11yError } from '@sa11y/format';

type FailureDetail = {
    error?: A11yError;
};
interface FailureMatcherDetail {
    error?: {
        matcherResult?: {
            a11yError?: A11yError;
        };
    };
}

type ErrorElement = {
    html: string;
    selectors: string;
    hierarchy: string;
    any: string;
    all: string;
    none: string;
};

type A11yViolation = {
    id: string;
    description: string;
    helpUrl: string;
    wcagCriteria: string;
    summary: string;
    errorElements: ErrorElement[];
};

const formatSpacing = '\t'.repeat(8);
const formatForAxeMessage = `\n${formatSpacing}\t\t`;

const axeMessages = {
    toSolveAny: `${formatForAxeMessage}- More Info: To solve the problem, you need to fix at least (1) of the following:\n`,
    toSolveFirst: `${formatForAxeMessage}- More Info: To solve the problem, you need to fix one of the following:\n`,
    toSolveSecond: `${formatForAxeMessage}- And fix the following:\n`,
};
/**
 * Create a test failure html elements array grouped by rule violation
 */
function createA11yErrorElements(errorElements: ErrorElement[]) {
    const a11yErrorElements: string[] = [];
    errorElements.forEach((errorElement, index) => {
        let errorMessage = `${formatSpacing}(${index + 1})  - HTML element : ${errorElement.html
            .replace(/&lt;/g, '<')
            .replace(/&gt;/, '>')}
                        - CSS selector(s) : ${errorElement.selectors.replace(/&gt;/, '>')}
                        - HTML Tag Hierarchy : ${errorElement.hierarchy}`;

        if (errorElement.any && errorElement.any.length > 0) {
            errorMessage += `${axeMessages.toSolveAny}${errorElement.any}`;
            if (errorElement.all && errorElement.all.length > 0) {
                errorMessage += `${axeMessages.toSolveSecond}${errorElement.all}`;
            }
            if (errorElement.none && errorElement.none.length > 0) {
                errorMessage += `${axeMessages.toSolveSecond}${errorElement.none}`;
            }
        } else if (errorElement.all && errorElement.all.length > 0) {
            errorMessage += `${axeMessages.toSolveFirst}${errorElement.all}`;
            if (errorElement.none && errorElement.none.length > 0) {
                errorMessage += `${axeMessages.toSolveSecond}${errorElement.none}`;
            }
        } else {
            if (errorElement.none && errorElement.none.length > 0) {
                errorMessage += `${axeMessages.toSolveFirst}${errorElement.none}`;
            }
        }
        a11yErrorElements.push(errorMessage);
    });

    return a11yErrorElements.join('\n');
}

/**
 * Create a test failure violation error message grouped by rule violation
 */
function createA11yRuleViolation(a11yRule: A11yViolation, ruleIndex: number) {
    return `(${ruleIndex}) [${a11yRule.id}] ${a11yRule.description}
            * Error element(s) : ${a11yRule.errorElements.length}\n${createA11yErrorElements(a11yRule.errorElements)}
            * Help:
                • Help URL: ${a11yRule.helpUrl}
                • WCAG Criteria: ${a11yRule.wcagCriteria}`;
}

/**
 * Create a test processA11yDetailsAndMessages violation error message grouped by rule violation
 */
function processA11yDetailsAndMessages(error: A11yError, a11yFailureMessages: string[]) {
    const a11yRuleViolations: { [key: string]: A11yViolation } = {};
    let a11yRuleViolationsCount = 0;
    let a11yErrorElementsCount = 0;
    error.a11yResults.forEach((a11yResult) => {
        a11yErrorElementsCount++;
        if (!(a11yRuleViolations as never)[a11yResult.wcag]) {
            a11yRuleViolationsCount++;
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
            a11yRuleViolations[a11yResult.wcag] = {
                id: a11yResult.id,
                description: a11yResult.description,
                helpUrl: a11yResult.helpUrl,
                wcagCriteria: a11yResult.wcag,
                summary: a11yResult.summary,
                errorElements: [],
            };
        }
        a11yRuleViolations[a11yResult.wcag].errorElements.push({
            html: a11yResult.html,
            selectors: a11yResult.selectors,
            hierarchy: a11yResult.ancestry,
            any: a11yResult.any,
            all: a11yResult.all,
            none: a11yResult.none,
        });
    });

    const a11yFailureMessage = `
    The test has failed the accessibility check. Accessibility Stacktrace/Issues:
    ${a11yErrorElementsCount} HTML elements have accessibility issue(s). ${a11yRuleViolationsCount} rules failed.

    ${Object.values(a11yRuleViolations)
        .map((a11yRuleViolation, index) => createA11yRuleViolation(a11yRuleViolation, index + 1))
        .join('\n')}


    For more info about automated accessibility testing: https://sfdc.co/a11y-test
    For tips on fixing accessibility bugs: https://sfdc.co/a11y
    For technical questions regarding Salesforce accessibility tools, contact our Sa11y team: http://sfdc.co/sa11y-users
    For guidance on accessibility related specifics, contact our A11y team: http://sfdc.co/tmp-a11y
    `;
    a11yFailureMessages.push(a11yFailureMessage);
}

/**
 * Convert any a11y errors from test failures into their own test suite, results
 */
function processA11yErrors(results: AggregatedResult, testSuite: TestResult, testResult: AssertionResult) {
    const a11yFailureDetails: FailureDetail[] = [];
    const a11yFailureMessages: string[] = [];
    let a11yErrorsExist = false;

    testResult.failureDetails.forEach((failure) => {
        let error = (failure as FailureDetail).error;
        // If using circus test runner https://github.com/facebook/jest/issues/11405#issuecomment-843549606
        // TODO (code cov): Add test data covering the case for circus test runner
        /* istanbul ignore next */
        if (error === undefined) error = failure as A11yError;
        if (error.name === A11yError.name) {
            a11yErrorsExist = true;
            a11yFailureDetails.push({ ...(failure as FailureDetail) } as FailureDetail);
            processA11yDetailsAndMessages(error, a11yFailureMessages);
        }
    });
    if (!a11yErrorsExist) {
        testSuite.numFailingTests -= 1;
        results.numFailedTests -= 1;
        if (testSuite.numFailingTests === 0) results.numFailedTestSuites -= 1;
    }
    testResult.failureDetails = [...a11yFailureDetails];
    testResult.failureMessages = [...a11yFailureMessages];
    testResult.status = a11yFailureDetails.length === 0 ? 'passed' : testResult.status;
}

function processA11yMatcherErrors(results: AggregatedResult, testSuite: TestResult, testResult: AssertionResult) {
    const a11yFailureMessages: string[] = [];

    testResult.failureDetails.forEach((failure) => {
        const error = (failure as FailureMatcherDetail)?.error?.matcherResult?.a11yError as A11yError;
        if (error !== undefined) {
            processA11yDetailsAndMessages(error, a11yFailureMessages);
            testResult.failureMessages = [...a11yFailureMessages];
        }
    });
}

/**
 * Custom results processor for a11y results grouping the violations. Only affects JSON results file output.
 * To be used with jest cli options --json --outputFile
 *  e.g. jest --json --outputFile jestResults.json --testResultsProcessor `node_modules/@sa11y/jest/dist/groupViolationResultsProcessor.js`
 * Ref: https://jestjs.io/docs/configuration#testresultsprocessor-string
 *  - Mapping of AggregatedResult to JSON format to https://github.com/facebook/jest/blob/master/packages/jest-test-result/src/formatTestResults.ts
 */
export function resultsProcessor(results: AggregatedResult): AggregatedResult {
    log(`Processing ${results.numTotalTests} tests ..`);
    results.testResults // suite results
        .filter((testSuite) => testSuite.numFailingTests > 0)
        .forEach((testSuite) => {
            testSuite.testResults // individual test results
                .filter((testResult) => testResult.status === 'failed')
                .forEach((testResult) => {
                    processA11yErrors(results, testSuite, testResult);
                });
        });

    return results;
}

export function resultsProcessorManualChecks(results: AggregatedResult): AggregatedResult {
    log(`Processing ${results.numTotalTests} tests ..`);
    results.testResults // suite results
        .filter((testSuite) => testSuite.numFailingTests > 0)
        .forEach((testSuite) => {
            testSuite.testResults // individual test results
                .filter((testResult) => testResult.status === 'failed')
                .forEach((testResult) => {
                    processA11yMatcherErrors(results, testSuite, testResult);
                });
        });

    return results;
}
// The processor must be a node module that exports a function
// Explicitly typing the exports for Node.js compatibility
const exportedFunctions = {
    resultsProcessor,
    resultsProcessorManualChecks,
};

module.exports = exportedFunctions;
