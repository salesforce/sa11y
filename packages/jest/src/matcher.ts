/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { matcherHint, printReceived } from 'jest-matcher-utils';
import { adaptA11yConfig } from './setup';
import { a11yCheckableContext, assertAccessible } from '@sa11y/assert';
import { A11yError, Options } from '@sa11y/format';
import { A11yConfig, extended } from '@sa11y/preset-rules';

// Type def for custom jest a11y matchers
// Ref: https://jestjs.io/docs/en/expect.html#expectextendmatchers
declare global {
    // eslint-disable-next-line @typescript-eslint/no-namespace
    namespace jest {
        interface Matchers<R> {
            toBeAccessible(config?: A11yConfig): Promise<CustomMatcherResult>;
        }
    }
}

const matcherHintMsg = `expected document to have no accessibility violations but found`;
const expectedMsg = `0 issues`;
const FormatOptions: Options = {
    a11yViolationIndicator: 'â­•',
    helpUrlIndicator: 'ðŸ”—',
    highlighter: printReceived,
};

/**
 * Jest expect matcher to check DOM for accessibility issues
 * @param received - DOM or HTML Element to be tested for accessibility. Defaults to current DOM.
 * @param config - A11yConfig to be used to test for accessibility. Defaults to extended.
 */
export async function toBeAccessible(
    received: a11yCheckableContext = document,
    config: A11yConfig = extended
): Promise<jest.CustomMatcherResult> {
    let isAccessible = true;
    let a11yError: A11yError;
    let receivedMsg = expectedMsg;

    // TODO (Improvement): Can we detect if this is invoked async and error if not ?
    try {
        await assertAccessible(received, adaptA11yConfig(config));
    } catch (e) {
        if (e instanceof A11yError) {
            a11yError = e;
            isAccessible = false;
            receivedMsg = `${a11yError.length} issues`;
        } else {
            throw e;
        }
    }

    return {
        pass: isAccessible,
        message: (): string =>
            matcherHint(
                `${toBeAccessible.name}: ${matcherHintMsg} ${receivedMsg}: \n\n ${a11yError.format(FormatOptions)}`,
                `${a11yError.length} issues`,
                expectedMsg
            ),
    };
}
