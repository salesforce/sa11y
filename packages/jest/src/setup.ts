/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { toBeAccessible } from './matcher';
import { A11yConfig } from '@sa11y/preset-rules';

/**
 * Register accessibility helpers toBeAccessible as jest matchers
 */
export function registerSa11yMatcher(): void {
    // Ref: https://github.com/jest-community/jest-extended
    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore error TS2339: Property 'expect' does not exist on type 'Global'.
    const jestExpect = global.expect;
    if (jestExpect !== undefined) {
        jestExpect.extend({ toBeAccessible });
    } else {
        throw new Error(
            "Unable to find Jest's global expect." +
                '\nPlease check you have added @sa11y/jest correctly to your jest configuration.' +
                '\nSee https://github.com/salesforce/sa11y/tree/master/packages/jest#readme for help.'
        );
    }
}

/**
 * Customize sa11y preset rules specific to JSDOM
 */
export function adaptA11yConfig(config: A11yConfig): A11yConfig {
    // TODO: Is it worth checking if we are running in jsdom before modifying config ?
    //  Ref: https://github.com/jsdom/jsdom/issues/1537#issuecomment-229405327
    // const runningInJSDOM = navigator.userAgent.includes('Node.js') || navigator.userAgent.includes('jsdom');
    // if (!runningInJSDOM) return config;
    return {
        ...config,
        rules: {
            // Disable color-contrast as it is doesn't work for JSDOM and might affect performance
            //  https://github.com/dequelabs/axe-core/issues/595
            //  https://github.com/dequelabs/axe-core/blob/develop/doc/examples/jsdom/test/a11y.js
            'color-contrast': { enabled: false },
        },
    };
}
