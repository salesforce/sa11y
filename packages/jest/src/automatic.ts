/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { AxeResults, log, useCustomRules } from '@sa11y/common';
import { getViolationsJSDOM } from '@sa11y/assert';
import { A11yError, exceptionListFilterSelectorKeywords } from '@sa11y/format';
import { isTestUsingFakeTimer } from './matcher';
import { expect } from '@jest/globals';
import { adaptA11yConfig, adaptA11yConfigCustomRules } from './setup';
import { defaultRuleset } from '@sa11y/preset-rules';

/**
 * Options for Automatic checks to be passed to {@link registerSa11yAutomaticChecks}
 */
export type AutoCheckOpts = {
    runAfterEach?: boolean;
    cleanupAfterEach?: boolean;
    consolidateResults?: boolean;
    // TODO (feat): add support for optional exclusion of selected tests
    // excludeTests?: string[];
    // List of test file paths (as regex) to filter for automatic checks
    filesFilter?: string[];
    runDOMMutationObserver?: boolean;
};

/**
 * Default options when {@link registerSa11yAutomaticChecks} is invoked
 */
const defaultAutoCheckOpts: AutoCheckOpts = {
    runAfterEach: true,
    cleanupAfterEach: true,
    consolidateResults: true,
    filesFilter: [],
};

let originalDocumentBodyHtml: string | null = null;

let mutatedNodes: string[] = [];

export const setOriginalDocumentBodyHtml = (bodyHtml: string | null) => {
    originalDocumentBodyHtml = bodyHtml ?? null;
};

export const getOriginalDocumentBodyHtml = () => {
    return originalDocumentBodyHtml;
};

/**
 * Check if current test file needs to be skipped based on any provided filter
 */
export function skipTest(testPath: string | undefined, filesFilter?: string[]): boolean {
    if (!testPath || !filesFilter || !(filesFilter?.length > 0)) return false;
    const skipTest = filesFilter.some((fileName) => testPath.toLowerCase().includes(fileName.toLowerCase()));

    if (skipTest) {
        log(
            `Skipping automatic accessibility check on ${testPath} as it matches given files filter: ${filesFilter.toString()}`
        );
    }
    return skipTest;
}

/**
 * Run accessibility check on each element node in the body using {@link toBeAccessible}
 * @param opts - Options for automatic checks {@link AutoCheckOpts}
 */
export async function automaticCheck(opts: AutoCheckOpts = defaultAutoCheckOpts): Promise<void> {
    if (skipTest(expect.getState().testPath, opts.filesFilter)) return;

    // Skip automatic check if test is using fake timer as it would result in timeout
    if (isTestUsingFakeTimer()) {
        log('Skipping automatic accessibility check as Jest fake timer is in use.');
        return;
    }

    let violations: AxeResults = [];
    const currentDocumentHtml = document.body.innerHTML;
    if (originalDocumentBodyHtml) {
        document.body.innerHTML = originalDocumentBodyHtml;
    }
    // Create a DOM walker filtering only elements (skipping text, comment nodes etc)
    const walker = document.createTreeWalker(document.body, NodeFilter.SHOW_ELEMENT);
    let currNode = walker.firstChild();
    const customRules = useCustomRules();
    try {
        if (!opts.runDOMMutationObserver) {
            while (currNode !== null) {
                // TODO (spike): Use a logger lib with log levels selectable at runtime
                // console.log(
                //     `â™¿ [DEBUG] Automatically checking a11y of ${currNode.nodeName}
                //      for test "${expect.getState().currentTestName}"
                //      : ${testPath}`
                // );
                // W-10004832 - Exclude descendancy based rules from automatic checks
                if (customRules.length === 0)
                    violations.push(...(await getViolationsJSDOM(currNode, adaptA11yConfig(defaultRuleset))));
                else
                    violations.push(
                        ...(await getViolationsJSDOM(currNode, adaptA11yConfigCustomRules(defaultRuleset, customRules)))
                    );
                currNode = walker.nextSibling();
            }
        } else {
            // loop mutated nodes
            for (let i = 0; i < mutatedNodes.length; i++) {
                if (mutatedNodes[i]) {
                    document.body.innerHTML = mutatedNodes[i];
                    const walker = document.createTreeWalker(document.body, NodeFilter.SHOW_ELEMENT);
                    let currNode = walker.firstChild();
                    while (currNode !== null) {
                        if (customRules.length === 0)
                            violations.push(...(await getViolationsJSDOM(currNode, adaptA11yConfig(defaultRuleset))));
                        else
                            violations.push(
                                ...(await getViolationsJSDOM(
                                    currNode,
                                    adaptA11yConfigCustomRules(defaultRuleset, customRules)
                                ))
                            );
                        currNode = walker.nextSibling();
                    }
                }
            }
        }
    } finally {
        if (opts.runDOMMutationObserver) {
            mutatedNodes = [];
        }
        setOriginalDocumentBodyHtml(null);
        document.body.innerHTML = currentDocumentHtml;
        if (opts.cleanupAfterEach) document.body.innerHTML = ''; // remove non-element nodes
        // TODO (spike): Disable stack trace for automatic checks.
        //  Will this affect all errors globally?
        // Error.stackTraceLimit = 0;
        if (process.env.SELECTOR_FILTER_KEYWORDS) {
            violations = exceptionListFilterSelectorKeywords(
                violations,
                process.env.SELECTOR_FILTER_KEYWORDS.split(',')
            );
        }
        A11yError.checkAndThrow(violations, { deduplicate: opts.consolidateResults });
    }
}

function observerCallback(mutations: MutationRecord[], _observer: MutationObserver) {
    for (const mutation of mutations) {
        mutation.addedNodes.forEach((node) => {
            if (node?.parentElement?.innerHTML) {
                mutatedNodes.push(node.parentElement.innerHTML);
            }
        });
    }
}

// https://developer.mozilla.org/en-US/docs/Web/API/MutationObserverInit
const observerOptions: MutationObserverInit = {
    subtree: true, // extend monitoring to the entire subtree of nodes rooted at target
    childList: true, // monitor target node for addition/removal of child nodes
    attributes: true, // monitor changes to the value of attributes of nodes
    characterData: true, // monitor changes to the character data contained within nodes
};

/**
 * Register accessibility checks to be run automatically after each test
 * @param opts - Options for automatic checks {@link AutoCheckOpts}
 */
export function registerSa11yAutomaticChecks(opts: AutoCheckOpts = defaultAutoCheckOpts): void {
    if (opts.runAfterEach) {
        const observer = new MutationObserver(observerCallback);
        // TODO (fix): Make registration idempotent
        log('Registering sa11y checks to be run automatically after each test');

        beforeEach(() => {
            if (opts.runDOMMutationObserver) {
                observer.observe(document.body, observerOptions);
            }
        });

        afterEach(async () => {
            if (opts.runDOMMutationObserver) {
                observer.disconnect(); // stop mutation observer
            }
            await automaticCheck(opts);
        });
    }
}
