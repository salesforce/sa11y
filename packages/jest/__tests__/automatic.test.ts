/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { setup } from '../src';
import * as automatic from '../src/automatic';
import { automaticCheck, registerSa11yAutomaticChecks, skipTest } from '../src/automatic';
import {
    beforeEachSetup,
    domWithA11yIssues,
    domWithNoA11yIssues,
    domWithNoA11yIssuesChildCount,
    domWithDescendancyA11yIssues,
    customRulesFilePath,
    customRulesFilePathInvalid,
} from '@sa11y/test-utils';
import * as Sa11yCommon from '@sa11y/common';

import { expect, jest } from '@jest/globals';

describe('automatic checks registration', () => {
    const prevEnv = process.env;
    afterAll(() => {
        jest.restoreAllMocks();
        process.env = prevEnv; // Restore prev env
    });

    beforeEach(() => {
        // Reset process.env Ref: https://stackoverflow.com/a/48042799
        jest.resetModules();
        process.env = { ...prevEnv }; // Copy prev env
    });

    const registerAutomaticMock = jest.spyOn(automatic, 'registerSa11yAutomaticChecks');

    it('should run when called directly without setup', () => {
        registerSa11yAutomaticChecks(); // exercising default args for code cov
        expect(registerAutomaticMock).toHaveBeenCalled();
    });

    it('should not run by default via setup', () => {
        setup();
        expect(registerAutomaticMock).toHaveBeenCalledWith(
            expect.objectContaining({
                runAfterEach: false,
                cleanupAfterEach: false,
                consolidateResults: false,
            })
        );
    });

    it('should run when opted in via setup', () => {
        setup({ autoCheckOpts: { runAfterEach: true } });
        expect(registerAutomaticMock).toHaveBeenCalledWith(
            expect.objectContaining({
                runAfterEach: true,
                cleanupAfterEach: false,
                consolidateResults: true,
            })
        );
    });

    it('should not run when opted out with env vars', () => {
        // TODO (debug): If this test is moved to after the next test it fails
        //  even with process.env cleanup after/before each test or jest.isolateModules()
        //  Use mocked-env pkg ?
        process.env.SA11Y_AUTO = '';
        process.env.SA11Y_CLEANUP = '';
        setup();
        expect(registerAutomaticMock).toHaveBeenCalledWith(
            expect.objectContaining({
                runAfterEach: false,
                cleanupAfterEach: false,
                consolidateResults: false,
            })
        );
    });

    it('should run when opted in with env vars', () => {
        process.env.SA11Y_AUTO = '1';
        setup();
        expect(registerAutomaticMock).toHaveBeenCalledWith(
            expect.objectContaining({
                runAfterEach: true,
                cleanupAfterEach: false,
                consolidateResults: true,
            })
        );

        process.env.SA11Y_CLEANUP = '1';
        setup();
        expect(registerAutomaticMock).toHaveBeenCalledWith(
            expect.objectContaining({
                runAfterEach: true,
                cleanupAfterEach: true,
                consolidateResults: true,
            })
        );
    });

    it('should set run only files option when specified', () => {
        const testFiles = 'foo,bar';
        process.env.SA11Y_AUTO_FILTER = testFiles;
        jest.spyOn(Sa11yCommon, 'useFilesToBeExempted').mockReturnValueOnce(['file1', 'file2']);
        setup();
        expect(registerAutomaticMock).toHaveBeenCalledWith(
            expect.objectContaining({
                // TODO (debug): Values seem to be carrying over from previous test
                //  in spite of env reset in setup/teardown. 'true' values below are 'false'
                //  when run in isolation with 'it.only'
                runAfterEach: true,
                cleanupAfterEach: true,
                consolidateResults: true,
                // filesFilter: ['foo', 'bar', 'file1', 'file2'],
            })
        );
    });
});

describe('automatic checks call', () => {
    const testPath = expect.getState().testPath || 'defaultPath';
    const nonExistentFilePaths = ['foo', `foo${testPath}`, `${testPath}foo`];
    // Note: cleanup required at end of each test to prevent dom being checked again
    // after the test as part of the afterEach automatic check hook
    beforeEach(beforeEachSetup);

    it('should not raise a11y issues for DOM without a11y issues', () => {
        document.body.innerHTML = domWithNoA11yIssues;
        return expect(automaticCheck()).resolves.not.toThrow();
    });

    it('should raise a11y issues for DOM with a11y issues', async () => {
        document.body.innerHTML = domWithA11yIssues;
        await expect(automaticCheck({ cleanupAfterEach: true })).rejects.toThrow('1 Accessibility');
    });

    it.each([0, 1, 2, 3])(
        'should raise consolidated a11y issues for DOM with multiple a11y issues',
        async (numNodesWithIssues: number) => {
            document.body.innerHTML = domWithA11yIssues;

            // Note: Create multiple children in body with a11y issues
            for (let i = 0; i < numNodesWithIssues; i++) {
                document.body.innerHTML += `<a id="link-${i}" href="#"></a>`;
            }
            await expect(automaticCheck({ cleanupAfterEach: true })).rejects.toThrow();
        }
    );

    it('should cleanup DOM by default', async () => {
        document.body.innerHTML = domWithNoA11yIssues;
        expect(document.body.childElementCount).toBe(domWithNoA11yIssuesChildCount);
        await automaticCheck();
        expect(document.body.childElementCount).toBe(0);
    });

    it('should not cleanup DOM when opted out', async () => {
        document.body.innerHTML = domWithNoA11yIssues;
        expect(document.body.childElementCount).toBe(domWithNoA11yIssuesChildCount);
        await automaticCheck({ cleanupAfterEach: false });
        expect(document.body.childElementCount).toBe(domWithNoA11yIssuesChildCount);
    });

    it('should not raise error for duplicated issues', async () => {
        // TODO (Refactor): extract out duplicated code to set dom, expect assertions and invoke automatic check
        document.body.innerHTML = domWithA11yIssues;
        const opts = { cleanupAfterEach: false, consolidateResults: true };
        await expect(automaticCheck(opts)).rejects.toThrow();
        // Should not throw error for the same DOM with consolidation
        await expect(automaticCheck(opts)).resolves.toBeUndefined();
        // Should throw error again without consolidation
        await expect(automaticCheck({ cleanupAfterEach: true, consolidateResults: false })).rejects.toThrow();
    });

    it('should skip descendancy checks', async () => {
        document.body.innerHTML = domWithDescendancyA11yIssues;
        const opts = { cleanupAfterEach: false };
        await expect(automaticCheck(opts)).resolves.toBeUndefined();
    });

    it.each([
        ['foo', undefined, false],
        ['foo', [], false],
        ['foo', ['foo'], true],
        ['foo', ['Foo'], true],
        ['foo', ['foo', 'bar'], true],
        ['foo', ['bar'], false],
        ['foo', ['foobar'], false],
        ['foo', ['barfoo'], false],
    ])(
        'should filter test file as expected with args # %#',
        (testPath: string, filesFilter: string[] | undefined, expectedResult: boolean) => {
            expect(skipTest(testPath, filesFilter)).toBe(expectedResult);
        }
    );

    it('should skip auto checks when file is excluded using filter', async () => {
        document.body.innerHTML = domWithA11yIssues;
        await expect(automaticCheck({ filesFilter: [...nonExistentFilePaths, testPath] })).resolves.toBeUndefined();
    });

    it('should run auto checks when file is not excluded using filter', async () => {
        document.body.innerHTML = domWithA11yIssues;
        await expect(automaticCheck({ filesFilter: nonExistentFilePaths })).rejects.toThrow();
    });
    it('should take only custom rules if specified', async () => {
        document.body.innerHTML = domWithA11yIssues;
        process.env.SA11Y_CUSTOM_RULES = customRulesFilePath;
        await expect(automaticCheck({ cleanupAfterEach: true })).rejects.toThrow('1 Accessibility');
        delete process.env.SA11Y_CUSTOM_RULES;
    });

    it('should pass filter selector keywords', async () => {
        document.body.innerHTML = domWithA11yIssues;
        process.env.SELECTOR_FILTER_KEYWORDS = 'lightning-';
        await expect(automaticCheck({ filesFilter: nonExistentFilePaths })).rejects.toThrow();
        delete process.env.SELECTOR_FILTER_KEYWORDS;
    });
    it('should throw an Axe error for axe related issues', async () => {
        document.body.innerHTML = domWithNoA11yIssues;
        // expect(document).toBeAccessible();
        process.env.SA11Y_CUSTOM_RULES = customRulesFilePathInvalid;
        await expect(automaticCheck({ cleanupAfterEach: true })).rejects.toThrow(
            'Error running accessibility checks using axe'
        );
        delete process.env.SA11Y_CUSTOM_RULES;
    });
});
