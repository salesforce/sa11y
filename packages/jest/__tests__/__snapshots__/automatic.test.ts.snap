// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`automatic checks call should not raise error for duplicated issues 1`] = `
"1 Accessibility issues found
 * (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name"
`;

exports[`automatic checks call should not raise error for duplicated issues 2`] = `
"1 Accessibility issues found
 * (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name"
`;

exports[`automatic checks call should raise a11y issues for DOM with a11y issues 1`] = `
"1 Accessibility issues found
 * (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name"
`;

exports[`automatic checks call should raise consolidated a11y issues for DOM with multiple a11y issues 1`] = `
"1 Accessibility issues found
 * (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name"
`;

exports[`automatic checks call should raise consolidated a11y issues for DOM with multiple a11y issues 2`] = `
"2 Accessibility issues found
 * (link-name) Links must have discernible text: #dom-with-issues > a
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name

* (link-name) Links must have discernible text: #link-0
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name"
`;

exports[`automatic checks call should raise consolidated a11y issues for DOM with multiple a11y issues 3`] = `
"3 Accessibility issues found
 * (link-name) Links must have discernible text: #dom-with-issues > a
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name

* (link-name) Links must have discernible text: #link-0
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name

* (link-name) Links must have discernible text: #link-1
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name"
`;

exports[`automatic checks call should raise consolidated a11y issues for DOM with multiple a11y issues 4`] = `
"4 Accessibility issues found
 * (link-name) Links must have discernible text: #dom-with-issues > a
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name

* (link-name) Links must have discernible text: #link-0
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name

* (link-name) Links must have discernible text: #link-1
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name

* (link-name) Links must have discernible text: #link-2
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name"
`;

exports[`automatic checks call should run auto checks when file is specified in run only option 1`] = `
"1 Accessibility issues found
 * (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name"
`;
